# IAM - Avaliação 01
**Disciplina:** Introdução ao Aprendizado de Máquina (ENGG67)  
**Professor:** Antônio C. L. Fernandes Jr.  
**Aluno(s):** Giovane Santos de Santana, Heverton Silva dos Reis  
**Data de Entrega:** 22/04/2025

---

### Conjunto de Dados: Breast Cancer Wisconsin (Diagnostic)
[Link para o dataset](https://archive.ics.uci.edu/dataset/17/breast+cancer+wisconsin+diagnostic)

---



Antes de começar a análise, é preciso carregar o dataset, através de importação direta, como indicado pela documentação do repositório:

In [ ]:
pip install ucimlrepo

In [ ]:
from ucimlrepo import fetch_ucirepo

# fetch dataset
breast_cancer_wisconsin_diagnostic = fetch_ucirepo(id=17)

# data (as pandas dataframes)
df = breast_cancer_wisconsin_diagnostic.data.features
dt = breast_cancer_wisconsin_diagnostic.data.targets

In [ ]:
import pandas as pd

### a) O conjunto de dados é univariado ou multivariado? Descreva a quantidade de atributos preditivos e atributos alvo.

Para avaliar se o conjunto de dados em análise são **univariados** ou **multivariados** é preciso expandir suas variáveis:

In [ ]:
# variable information
print(breast_cancer_wisconsin_diagnostic.variables)

                  name     role         type demographic description units  \
0                   ID       ID  Categorical        None        None  None   
1            Diagnosis   Target  Categorical        None        None  None   
2              radius1  Feature   Continuous        None        None  None   
3             texture1  Feature   Continuous        None        None  None   
4           perimeter1  Feature   Continuous        None        None  None   
5                area1  Feature   Continuous        None        None  None   
6          smoothness1  Feature   Continuous        None        None  None   
7         compactness1  Feature   Continuous        None        None  None   
8           concavity1  Feature   Continuous        None        None  None   
9      concave_points1  Feature   Continuous        None        None  None   
10           symmetry1  Feature   Continuous        None        None  None   
11  fractal_dimension1  Feature   Continuous        None        

Através dessa expansão, percebe-se que o conjunto de dados é **multivariado**, pois entre suas variáveis verificou-se a ocorrência de até **30 atributos preditivos** (identificados no dataset com o papel de *Feature*). Dentre os outros atributos, o atributo que possuiu o papel de **alvo** (*Target*) é o atributo **Diagnosis**.

### b)  Caracterize os dados quanto ao Tipo e Escala.

Para definir o **Tipo** e **Escala** dos dados, pode-se começar verificando os tipos dos **atributos preditivos** através do comando *dtypes* da biblioteca *Pandas*:

In [ ]:
# Ver os tipos de dados diretamente do DataFrame
df.dtypes

radius1               float64
texture1              float64
perimeter1            float64
area1                 float64
smoothness1           float64
compactness1          float64
concavity1            float64
concave_points1       float64
symmetry1             float64
fractal_dimension1    float64
radius2               float64
texture2              float64
perimeter2            float64
area2                 float64
smoothness2           float64
compactness2          float64
concavity2            float64
concave_points2       float64
symmetry2             float64
fractal_dimension2    float64
radius3               float64
texture3              float64
perimeter3            float64
area3                 float64
smoothness3           float64
compactness3          float64
concavity3            float64
concave_points3       float64
symmetry3             float64
fractal_dimension3    float64
dtype: object

Logo, é possível concluir que os **atributos preditivos** são do tipo **numérico contínuo**.

Já para o **atributo alvo**, optou-se por plotar as primeiras linhas do DataFrame através do comando *head()*, que revelou que esse atributo é do tipo **categórico**.

In [ ]:
# Ver os dados das primeiras linhas do DataFrame
dt.head()

Diagnosis
0         M
1         M
2         M
3         M
4         M

Quanto a escala, para o atributo alvo, é imedato perceber que se tratam de valores **nominais**. Já para os atributos preditivos, é preciso prestar atenção ao contexto e perceber que eles fazem alusão a atributos geométricos como área, perímetro, concavidade, simetria, entre outros. Ou seja, esses atributos carregam significado absoluto, portanto tais atributos são da escala **racional**.

### c) Verifique a necessidade de balanceamento dos dados e caracterize estatisticamente os dados (calcule, para cada atributo, suas estatísticas univariadas pertinentes dentre frequência, moda, média, mediana, desvio-padrão, Q1, Q3, obliquidade e curtose).

Para verificar se o conjunto precisa de um balancemanto, pode-se começar analisando a distribuição das classes de *Diagnosis*, o **atributo alvo**.

In [ ]:
# Ver distribuição das classes do diagnóstico
dt['Diagnosis'].value_counts()

Diagnosis
B    357
M    212
Name: count, dtype: int64

Tem-se que existem **357 amostras** da classe **B (Benigno)** e **212 amostras** da classe **M (Maligino)**, num **total de 569 amostras**. Ao verificar a proporção das amostras, observou-se que:
<br>
<br>
<br>
$Classe B = \frac{357}{569} \times 100 = 62.7 \%$ e $Classe A = \frac{212}{569} \times 100 = 37.3 \%$

In [ ]:
# Selecionar apenas os atributos preditivos

# Estatísticas básicas
descricao = pd.DataFrame()
descricao['Média'] = df.mean()
descricao['Mediana'] = df.median()
descricao['Moda'] = df.mode().iloc[0]
descricao['Desvio Padrão'] = df.std()
descricao['Q1 (25%)'] = df.quantile(0.25)
descricao['Q3 (75%)'] = df.quantile(0.75)
descricao['Obliquidade'] = df.skew()
descricao['Curtose'] = df.kurt()

# Exibir as variáveis
descricao.head(30)

Média     Mediana        Moda  Desvio Padrão  \
radius1              14.127292   13.370000   12.340000       3.524049   
texture1             19.289649   18.840000   14.930000       4.301036   
perimeter1           91.969033   86.240000   82.610000      24.298981   
area1               654.889104  551.100000  512.200000     351.914129   
smoothness1           0.096360    0.095870    0.100700       0.014064   
compactness1          0.104341    0.092630    0.114700       0.052813   
concavity1            0.088799    0.061540    0.000000       0.079720   
concave_points1       0.048919    0.033500    0.000000       0.038803   
symmetry1             0.181162    0.179200    0.160100       0.027414   
fractal_dimension1    0.062798    0.061540    0.056670       0.007060   
radius2               0.405172    0.324200    0.220400       0.277313   
texture2              1.216853    1.108000    0.856100       0.551648   
perimeter2            2.866059    2.287000    1.778000       2.021855   
area2                40.337079   24.530000   16.640000      45.491006   
smoothness2           0.007041    0.006380    0.005080       0.003003   
compactness2          0.025478    0.020450    0.011040       0.017908   
concavity2            0.031894    0.025890    0.000000       0.030186   
concave_points2       0.011796    0.010930    0.000000       0.006170   
symmetry2             0.020542    0.018730    0.013440       0.008266   
fractal_dimension2    0.003795    0.003187    0.001784       0.002646   
radius3              16.269190   14.970000   12.360000       4.833242   
texture3             25.677223   25.410000   17.700000       6.146258   
perimeter3          107.261213   97.660000  101.700000      33.602542   
area3               880.583128  686.500000  284.400000     569.356993   
smoothness3           0.132369    0.131300    0.121600       0.022832   
compactness3          0.254265    0.211900    0.148600       0.157336   
concavity3            0.272188    0.226700    0.000000       0.208624   
concave_points3       0.114606    0.099930    0.000000       0.065732   
symmetry3             0.290076    0.282200    0.222600       0.061867   
fractal_dimension3    0.083946    0.080040    0.074270       0.018061   

                      Q1 (25%)     Q3 (75%)  Obliquidade    Curtose  
radius1              11.700000    15.780000     0.942380   0.845522  
texture1             16.170000    21.800000     0.650450   0.758319  
perimeter1           75.170000   104.100000     0.990650   0.972214  
area1               420.300000   782.700000     1.645732   3.652303  
smoothness1           0.086370     0.105300     0.456324   0.855975  
compactness1          0.064920     0.130400     1.190123   1.650130  
concavity1            0.029560     0.130700     1.401180   1.998638  
concave_points1       0.020310     0.074000     1.171180   1.066556  
symmetry1             0.161900     0.195700     0.725609   1.287933  
fractal_dimension1    0.057700     0.066120     1.304489   3.005892  
radius2               0.232400     0.478900     3.088612  17.686726  
texture2              0.833900     1.474000     1.646444   5.349169  
perimeter2            1.606000     3.357000     3.443615  21.401905  
area2                17.850000    45.190000     5.447186  49.209077  
smoothness2           0.005169     0.008146     2.314450  10.469840  
compactness2          0.013080     0.032450     1.902221   5.106252  
concavity2            0.015090     0.042050     5.110463  48.861395  
concave_points2       0.007638     0.014710     1.444678   5.126302  
symmetry2             0.015160     0.023480     2.195133   7.896130  
fractal_dimension2    0.002248     0.004558     3.923969  26.280847  
radius3              13.010000    18.790000     1.103115   0.944090  
texture3             21.080000    29.720000     0.498321   0.224302  
perimeter3           84.110000   125.400000     1.128164   1.070150  
area3               515.300000  1084.000000     1.859373   4.396395  
smoothness3

### d) Analise os dados quanto à sua incompletude, inconsistências, redundâncias e presença de ruído

Uma estatégia que pode ser utilizada para avaliar o conjunto de dados quanto à sua **incompletude** é verificar a existência de **valores nulos** entre os atributos. Isso pode ser feito atraves do comando *isnull()* do pandas. Ao se fazer um **somatório desses valores**, se o valor for 0 para cada uma das colunas, então **não existe incompletude** para esses atributos. Começa-se com os atributos preditivos:


In [ ]:
# Verificar valores nulos em cada coluna
df.isnull().sum()

radius1               0
texture1              0
perimeter1            0
area1                 0
smoothness1           0
compactness1          0
concavity1            0
concave_points1       0
symmetry1             0
fractal_dimension1    0
radius2               0
texture2              0
perimeter2            0
area2                 0
smoothness2           0
compactness2          0
concavity2            0
concave_points2       0
symmetry2             0
fractal_dimension2    0
radius3               0
texture3              0
perimeter3            0
area3                 0
smoothness3           0
compactness3          0
concavity3            0
concave_points3       0
symmetry3             0
fractal_dimension3    0
dtype: int64

O mesmo pode ser feito para o atributo alvo:

In [ ]:
# Verificar valores nulos em cada coluna
dt.isnull().sum()

Diagnosis    0
dtype: int64

Assim, pode-se concluir que não existem incompletudes entre os valores do dataset.

Já para verificar o conjunto em relação às **incosistências**, pode-se partir da análise já realizada anteriormente sobre os **Tipo** e a **Escala** dos dados. Como visto, os atributos preditivos representam valores **numéricos contínuos e absolutos**, mais especificamente, tratam de **medidas físicas** as quais **não se aplicam valores negativos** (a exemplo de área, raio, concavidade, etc.). Logo, assim como na abordagem anterior, pode-se verificar a existência desses valores e calcular seu somatório. Novamente, busca-se que o valor seja 0, para que não hajam incosistências:

In [ ]:
# Verificar valores negativos dos atributos preditivos
(df < 0).sum()


radius1               0
texture1              0
perimeter1            0
area1                 0
smoothness1           0
compactness1          0
concavity1            0
concave_points1       0
symmetry1             0
fractal_dimension1    0
radius2               0
texture2              0
perimeter2            0
area2                 0
smoothness2           0
compactness2          0
concavity2            0
concave_points2       0
symmetry2             0
fractal_dimension2    0
radius3               0
texture3              0
perimeter3            0
area3                 0
smoothness3           0
compactness3          0
concavity3            0
concave_points3       0
symmetry3             0
fractal_dimension3    0
dtype: int64

Então, tem-se que os valores dos atributos preditivos são consistentes.

Quanto a redundânicia, uma forma de avaliar isso é possível aplicar uma **matriz de correlação**, e verificar os atributos que possuem uma correlação **próxima de 1**. Essa característica estatística pode indicar atributos que trazem informações semelhantes, e, que um deles poderia ser removido, a depender da análise.

In [ ]:
# Matriz de correlação
cor = df.corr().abs()

# Verificar redundância: pares com correlação alta (> 0.9)
redundantes = cor.where((cor > 0.9) & (cor < 1.0))
redundantes.dropna(how='all', axis=0).dropna(how='all', axis=1)

radius1  texture1  perimeter1     area1  concavity1  \
radius1               NaN       NaN    0.997855  0.987357         NaN   
texture1              NaN       NaN         NaN       NaN         NaN   
perimeter1       0.997855       NaN         NaN  0.986507         NaN   
area1            0.987357       NaN    0.986507       NaN         NaN   
concavity1            NaN       NaN         NaN       NaN         NaN   
concave_points1       NaN       NaN         NaN       NaN    0.921391   
radius2               NaN       NaN         NaN       NaN         NaN   
perimeter2            NaN       NaN         NaN       NaN         NaN   
area2                 NaN       NaN         NaN       NaN         NaN   
radius3          0.969539       NaN    0.969476  0.962746         NaN   
texture3              NaN  0.912045         NaN       NaN         NaN   
perimeter3       0.965137       NaN    0.970387  0.959120         NaN   
area3            0.941082       NaN    0.941550  0.959213         NaN   
concave_points3       NaN       NaN         NaN       NaN         NaN   

                 concave_points1   radius2  perimeter2     area2   radius3  \
radius1                      NaN       NaN         NaN       NaN  0.969539   
texture1                     NaN       NaN         NaN       NaN       NaN   
perimeter1                   NaN       NaN         NaN       NaN  0.969476   
area1                        NaN       NaN         NaN       NaN  0.962746   
concavity1              0.921391       NaN         NaN       NaN       NaN   
concave_points1              NaN       NaN         NaN       NaN       NaN   
radius2                      NaN       NaN    0.972794  0.951830       NaN   
perimeter2                   NaN  0.972794         NaN  0.937655       NaN   
area2                        NaN  0.951830    0.937655       NaN       NaN   
radius3                      NaN       NaN         NaN       NaN       NaN   
texture3                     NaN       NaN         NaN       NaN       NaN   
perimeter3                   NaN       NaN         NaN       NaN  0.993708   
area3                        NaN       NaN         NaN       NaN  0.984015   
concave_points3         0.910155       NaN         NaN       NaN       NaN   

                 texture3  perimeter3     area3  concave_points3  
radius1               NaN    0.965137  0.941082              NaN  
texture1         0.912045         NaN       NaN              NaN  
perimeter1            NaN    0.970387  0.941550              NaN  
area1                 NaN    0.959120  0.959213              NaN  
concavity1            NaN         NaN       NaN              NaN  
concave_points1       NaN         NaN       NaN         0.910155  
radius2               NaN         NaN       NaN              NaN  
perimeter2            NaN         NaN       NaN              NaN  
area2                 NaN         NaN       NaN              NaN  
radius3               NaN    0.993708  0.984015              NaN  
texture3              NaN         NaN       NaN              NaN  
perimeter3            NaN         NaN  0.977578              NaN  
area3                 NaN    0.977578       NaN              NaN  
concave_points3       NaN         NaN       NaN              NaN

Ao fazer isso com os atributos preditivos, como esperado, verificou-se que existem correlações altas entre atributos físicos, como **área, perímetro e raio**, por exemplo. Logo, alguns desses valores poderiam ser desprezados na análise de alguns modelos, como os de regressão linear por exemplo.

### e) Calcule estatísticas multivariadas de covariância e correlação para os atributos quantitativos presentes. Interprete os valores obtidos.

In [ ]:
df.corr()

radius1  texture1  perimeter1     area1  smoothness1  \
radius1             1.000000  0.323782    0.997855  0.987357     0.170581   
texture1            0.323782  1.000000    0.329533  0.321086    -0.023389   
perimeter1          0.997855  0.329533    1.000000  0.986507     0.207278   
area1               0.987357  0.321086    0.986507  1.000000     0.177028   
smoothness1         0.170581 -0.023389    0.207278  0.177028     1.000000   
compactness1        0.506124  0.236702    0.556936  0.498502     0.659123   
concavity1          0.676764  0.302418    0.716136  0.685983     0.521984   
concave_points1     0.822529  0.293464    0.850977  0.823269     0.553695   
symmetry1           0.147741  0.071401    0.183027  0.151293     0.557775   
fractal_dimension1 -0.311631 -0.076437   -0.261477 -0.283110     0.584792   
radius2             0.679090  0.275869    0.691765  0.732562     0.301467   
texture2           -0.097317  0.386358   -0.086761 -0.066280     0.068406   
perimeter2          0.674172  0.281673    0.693135  0.726628     0.296092   
area2               0.735864  0.259845    0.744983  0.800086     0.246552   
smoothness2        -0.222600  0.006614   -0.202694 -0.166777     0.332375   
compactness2        0.206000  0.191975    0.250744  0.212583     0.318943   
concavity2          0.194204  0.143293    0.228082  0.207660     0.248396   
concave_points2     0.376169  0.163851    0.407217  0.372320     0.380676   
symmetry2          -0.104321  0.009127   -0.081629 -0.072497     0.200774   
fractal_dimension2 -0.042641  0.054458   -0.005523 -0.019887     0.283607   
radius3             0.969539  0.352573    0.969476  0.962746     0.213120   
texture3            0.297008  0.912045    0.303038  0.287489     0.036072   
perimeter3          0.965137  0.358040    0.970387  0.959120     0.238853   
area3               0.941082  0.343546    0.941550  0.959213     0.206718   
smoothness3         0.119616  0.077503    0.150549  0.123523     0.805324   
compactness3        0.413463  0.277830    0.455774  0.390410     0.472468   
concavity3          0.526911  0.301025    0.563879  0.512606     0.434926   
concave_points3     0.744214  0.295316    0.771241  0.722017     0.503053   
symmetry3           0.163953  0.105008    0.189115  0.143570     0.394309   
fractal_dimension3  0.007066  0.119205    0.051019  0.003738     0.499316   

                    compactness1  concavity1  concave_points1  symmetry1  \
radius1                 0.506124    0.676764         0.822529   0.147741   
texture1                0.236702    0.302418         0.293464   0.071401   
perimeter1              0.556936    0.716136         0.850977   0.183027   
area1                   0.498502    0.685983         0.823269   0.151293   
smoothness1             0.659123    0.521984         0.553695   0.557775   
compactness1            1.000000    0.883121         0.831135   0.602641   
concavity1              0.883121    1.000000         0.921391   0.500667   
concave_points1         0.831135    0.921391         1.000000   0.462497   
symmetry1               0.602641    0.500667         0.462497   1.000000   
fractal_dimension1      0.565369    0.336783         0.166917   0.479921   
radius2                 0.497473    0.631925         0.698050   0.303379   
texture2                0.046205    0.076218         0.021480   0.128053   
perimeter2              0.548905    0.660391         0.710650   0.313893   
area2                   0.455653    0.617427         0.690299   0.223970   
smoothness2             0.135299    0.098564         0.027653   0.187321   
compactness2            0.738722    0.670279         0.490424   0.421659   
concavity2              0.570517    0.691270         0.439167   0.342627   
concave_points2         0.642262    0.683260         0.615634   0.393298   
symmetry2               0.229977    0.178009         0.095351   0.449137   
fractal_dimension2      0.507318    0.449301         0.257584   0.331786   
radius3                 0.535315    0